src/defect/DefectService.ts

```diff
@@ -0,0 +1,223 @@
+import moment from 'moment';
+import { Service } from 'typedi';
+import { ICustomDefect } from '../models';
+import { IVA_30, LOCATION_ENGLISH, LOCATION_WELSH } from '../models/Enums';
+import { IDefectChild } from '../models/IDefectChild';
+import { IDefectParent } from '../models/IDefectParent';
+import { IFlatDefect } from '../models/IFlatDefect';
+import { IItem } from '../models/IItem';
+import { TestResultService } from '../test-result/TestResultService';
+
+@Service()
+export class DefectService {
+	constructor(private testResultService: TestResultService) {}
+
+	/**
+	 * Formats the additional defects for IVA and MSVA test based on whether custom defects is populated
+	 * @param customDefects - the custom defects for the test
+	 */
+	public formatVehicleApprovalAdditionalDefects = (
+		customDefects: ICustomDefect[] | undefined
+	): ICustomDefect[] | undefined => {
+		const defaultCustomDefect: ICustomDefect = {
+			defectName: IVA_30.EMPTY_CUSTOM_DEFECTS,
+			defectNotes: '',
+		};
+		return customDefects && customDefects.length > 0 ? customDefects : [defaultCustomDefect];
+	};
+
+	/**
+	 * Returns a flattened array of every deficiency that only includes the key/value pairs required for certificate generation
+	 * @param defects - the array of defects from the api
+	 */
+	public flattenDefectsFromApi(defects: IDefectParent[]): IFlatDefect[] {
+		const flatDefects: IFlatDefect[] = [];
+		try {
+			// go through each defect in un-flattened array
+			defects.forEach((defect: IDefectParent) => {
```

Comment:
I think using a flat map might be more useful here instead of the nested for loops, so would have a flatmap to get the im number, im descirptin, and desc welsh, then another flatmap to go thorgugh the items and destructure the item number etc... then finally return the deficiciencies map and directly return an array, instead of having to make an array and push to it each time

Reply:
Same as above comment.